File: original/Code/Commando/cspingrequestevent.cpp
#include "global.h"
#include "cspingrequestevent.h"


#include "networkobjectfactory.h"
#include "cnetwork.h"
#include "networkobjectmgr.h"
#include "apppackettypes.h"
#include "scpingresponseevent.h"

DECLARE_NETWORKOBJECT_FACTORY(cCsPingRequestEvent, NETCLASSID_CSPINGREQUESTEVENT);

//-----------------------------------------------------------------------------
cCsPingRequestEvent::cCsPingRequestEvent(void)
{
	SenderId		= -1;
	PingNumber	= -1;

	Set_App_Packet_Type(APPPACKETTYPE_CSPINGREQUESTEVENT);
}

//-----------------------------------------------------------------------------
void
cCsPingRequestEvent::Init(int ping_number)
{
	//

	SenderId		= cNetwork::Get_My_Id();
	PingNumber	= ping_number;

	Set_Network_ID(NetworkObjectMgrClass::Get_New_Client_ID());

	Set_Object_Dirty_Bit(0, BIT_CREATION, true);
}

//-----------------------------------------------------------------------------
void
cCsPingRequestEvent::Act(void)
{

	cScPingResponseEvent * p_event = new cScPingResponseEvent;
	p_event->Init(SenderId, PingNumber);

	Set_Delete_Pending();
}

//-----------------------------------------------------------------------------
void
cCsPingRequestEvent::Export_Creation(BitStreamClass & packet)
{
	//

	cNetEvent::Export_Creation(packet);

	packet.Add(SenderId);
	packet.Add(PingNumber);

	Set_Delete_Pending();
}

//-----------------------------------------------------------------------------
void
cCsPingRequestEvent::Import_Creation(BitStreamClass & packet)
{

	cNetEvent::Import_Creation(packet);

	packet.Get(SenderId);
	packet.Get(PingNumber);

	Act();
}

File: kotlin-server/server/src/main/kotlin/ccr/server/net/CsPingRequestEvent.kt
package ccr.server.net

import ccr.net.bitstream.BitStream
import ccr.server.Network

// C++: cCsPingRequestEvent — networkClassId = NETCLASSID_CSPINGREQUESTEVENT = 1032
// Client→Server ping request event used to measure round-trip latency.
// Wire format (BIT_CREATION):
//   SenderId (int)
//   PingNumber (int)
class CsPingRequestEvent(
    var senderId: Int = 0,
    var pingNumber: Int = 0,
) : NetEvent() {
    override val networkClassId: Int = 1032

    override fun exportCreation(packet: BitStream) {
        packet.addInt(senderId)
        packet.addInt(pingNumber)
    }

    override fun importCreation(packet: BitStream) {
        senderId = packet.getInt()
        pingNumber = packet.getInt()
        actIfWiredUp()
    }

    override fun act() {
        println("[GAME] CSPINGREQUESTEVENT from senderId=$senderId pingNumber=$pingNumber → ScPingResponseEvent")
        val host = network.connectionManager.getHost(senderId) ?: run { setDeletePending(); return }
        val response = ScPingResponseEvent(pingNumber)
        network.serverSendPacket(host) { bs -> NetworkObjectPacketWriter.writeCreation(bs, response, response.networkId) }
        setDeletePending()
    }
}
